applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,741 @@
+		pos += sizeof(ble_params.eval_duration);
+
+		params->eval_keepout_duration =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(ble_params.eval_keepout_duration);
+
+		params->eval_success_threshold =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(ble_params.eval_success_threshold);
+	}
+	if (qos_wifi_params != NULL) {
+		pos = 0;
+
+		params->wifi_rating_inc =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(wifi_params.wifi_rating_inc);
+
+		params->wifi_present_threshold =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(wifi_params.wifi_present_threshold);
+
+		params->wifi_active_threshold =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(wifi_params.wifi_active_threshold);
+	}
+
+	k_mutex_lock(&data_access_mutex, K_FOREVER);
+	if (new_params != NULL) {
+		k_free(new_params);
+	}
+	new_params = params;
+	k_mutex_unlock(&data_access_mutex);
+}
+
+static void update_blacklist(const u8_t *blacklist)
+{
+	atomic_set(&new_blacklist, sys_get_le16(blacklist));
+}
+
+static int hci_chmap_cmd_send(const u8_t *chmap)
+{
+	struct bt_hci_cp_le_set_host_chan_classif *cp;
+	struct net_buf *buf;
+
+	__ASSERT_NO_MSG(chmap != NULL);
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_HOST_CHAN_CLASSIF,
+				sizeof(*cp));
+	if (!buf) {
+		return -1;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	if (!cp) {
+		return -1;
+	}
+
+	memcpy(cp->ch_map, chmap, sizeof(cp->ch_map));
+
+	return bt_hci_cmd_send_sync(
+			BT_HCI_OP_LE_SET_HOST_CHAN_CLASSIF,
+			buf,
+			NULL);
+}
+
+static bool on_vs_evt(struct net_buf_simple *buf)
+{
+	u8_t *subevent_code;
+	hci_vs_evt_qos_conn_event_report_t *evt;
+
+	subevent_code = net_buf_simple_pull_mem(
+						buf,
+						sizeof(*subevent_code));
+
+	switch (*subevent_code) {
+	case HCI_VS_SUBEVENT_CODE_QOS_CONN_EVENT_REPORT:
+		if (atomic_get(&processing) == true) {
+			/* Cheaper to skip this update */
+			/* instead of using locks */
+			return true;
+		}
+
+		evt = (void *)buf->data;
+
+		chmap_filter_crc_update(
+			chmap_inst,
+			evt->channel_index,
+			evt->crc_ok_count,
+			evt->crc_error_count);
+		return true;
+	default:
+		return false;
+	}
+}
+
+static void enable_qos_reporting(void)
+{
+	int err;
+	struct net_buf *buf;
+
+	err = bt_hci_register_vnd_evt_cb(on_vs_evt);
+	if (err) {
+		LOG_ERR("Failed to register HCI VS callback");
+		return;
+	}
+
+	hci_vs_cmd_qos_conn_event_report_enable_t *cmd_enable;
+
+	buf = bt_hci_cmd_create(HCI_VS_OPCODE_CMD_QOS_CONN_EVENT_REPORT_ENABLE,
+							sizeof(*cmd_enable));
+	if (!buf) {
+		LOG_ERR("Failed to enable HCI VS QoS");
+		return;
+	}
+
+	cmd_enable = net_buf_add(buf, sizeof(*cmd_enable));
+	cmd_enable->enable = 1;
+
+	err = bt_hci_cmd_send(
+			HCI_VS_OPCODE_CMD_QOS_CONN_EVENT_REPORT_ENABLE,
+			buf);
+	if (err) {
+		LOG_ERR("Failed to enable HCI VS QoS");
+		return;
+	}
+}
+
+static void handle_config_event(const struct config_event *event)
+{
+	switch (OPT_FIELD_GET(event->id)) {
+	case QOS_OPT_BLACKLIST:
+		__ASSERT_NO_MSG(
```

Comment:
assert is not enough - since data comes from outside world it musts be properly validated